File: java/designpattern/creational/SingletonPattern.java
/**
 * Singleton Design Pattern
 * 
 * Intent: Ensure a class has only one instance and provide a global point of access to it.
 * 
 * Use cases:
 * 1. Database connections
 * 2. Configuration settings
 * 3. Logger instances
 * 4. Cache managers
 * 5. Thread pools
 * 
 * Implementation approaches:
 * 1. Eager initialization
 * 2. Lazy initialization with synchronization
 * 3. Double-checked locking
 * 4. Bill Pugh Singleton (Static inner class)
 * 5. Enum Singleton
 */

// 1. Eager Initialization Singleton
class EagerSingleton {
    // Private static instance created when class is loaded
    private static final EagerSingleton instance = new EagerSingleton();
    
    // Private constructor to prevent instantiation
    private EagerSingleton() {
        System.out.println("EagerSingleton instance created");
    }
    
    // Public method to get the instance
    public static EagerSingleton getInstance() {
        return instance;
    }
    
    public void doSomething() {
        System.out.println("EagerSingleton: Doing something...");
    }
}

// 2. Lazy Initialization Singleton with Synchronization
class LazySingleton {
    private static LazySingleton instance;
    
    private LazySingleton() {
        System.out.println("LazySingleton instance created");
    }
    
    // Synchronized method to prevent multiple threads from creating instances
    public static synchronized LazySingleton getInstance() {
        if (instance == null) {
            instance = new LazySingleton();
        }
        return instance;
    }
    
    public void doSomething() {
        System.out.println("LazySingleton: Doing something...");
    }
}

// 3. Double-Checked Locking Singleton
class DoubleCheckedSingleton {
    // Volatile keyword ensures multiple threads handle the instance variable correctly
    private static volatile DoubleCheckedSingleton instance;
    
    private DoubleCheckedSingleton() {
        System.out.println("DoubleCheckedSingleton instance created");
    }
    
    public static DoubleCheckedSingleton getInstance() {
        if (instance == null) {
            synchronized (DoubleCheckedSingleton.class) {
                if (instance == null) {
                    instance = new DoubleCheckedSingleton();
                }
            }
        }
        return instance;
    }
    
    public void doSomething() {
        System.out.println("DoubleCheckedSingleton: Doing something...");
    }
}

// 4. Bill Pugh Singleton (Static Inner Class)
class BillPughSingleton {
    private BillPughSingleton() {
        System.out.println("BillPughSingleton instance created");
    }
    
    // Private inner class that holds the singleton instance
    private static class SingletonHelper {
        private static final BillPughSingleton INSTANCE = new BillPughSingleton();
    }
    
    public static BillPughSingleton getInstance() {
        return SingletonHelper.INSTANCE;
    }
    
    public void doSomething() {
        System.out.println("BillPughSingleton: Doing something...");
    }
}

// 5. Enum Singleton (Thread-safe and serialization-safe)
enum EnumSingleton {
    INSTANCE;
    
    private EnumSingleton() {
        System.out.println("EnumSingleton instance created");
    }
    
    public void doSomething() {
        System.out.println("EnumSingleton: Doing something...");
    }
}

// Example usage with a Database Connection
class DatabaseConnection {
    private static DatabaseConnection instance;
    private String connectionString;
    
    private DatabaseConnection() {
        // Simulate database connection initialization
        this.connectionString = "jdbc:mysql://localhost:3306/mydb";
        System.out.println("Database connection established");
    }
    
    public static synchronized DatabaseConnection getInstance() {
        if (instance == null) {
            instance = new DatabaseConnection();
        }
        return instance;
    }
    
    public void query(String sql) {
        System.out.println("Executing query: " + sql);
    }
    
    public void close() {
        System.out.println("Closing database connection");
    }
}

public class SingletonPattern {
    public static void main(String[] args) {
        // Test all singleton implementations
        System.out.println("Testing Singleton Pattern Implementations\n");
        
        // 1. Eager Singleton
        System.out.println("1. Testing Eager Singleton:");
        EagerSingleton eager1 = EagerSingleton.getInstance();
        EagerSingleton eager2 = EagerSingleton.getInstance();
        System.out.println("Are instances same? " + (eager1 == eager2));
        eager1.doSomething();
        
        // 2. Lazy Singleton
        System.out.println("\n2. Testing Lazy Singleton:");
        LazySingleton lazy1 = LazySingleton.getInstance();
        LazySingleton lazy2 = LazySingleton.getInstance();
        System.out.println("Are instances same? " + (lazy1 == lazy2));
        lazy1.doSomething();
        
        // 3. Double-Checked Singleton
        System.out.println("\n3. Testing Double-Checked Singleton:");
        DoubleCheckedSingleton dcs1 = DoubleCheckedSingleton.getInstance();
        DoubleCheckedSingleton dcs2 = DoubleCheckedSingleton.getInstance();
        System.out.println("Are instances same? " + (dcs1 == dcs2));
        dcs1.doSomething();
        
        // 4. Bill Pugh Singleton
        System.out.println("\n4. Testing Bill Pugh Singleton:");
        BillPughSingleton bps1 = BillPughSingleton.getInstance();
        BillPughSingleton bps2 = BillPughSingleton.getInstance();
        System.out.println("Are instances same? " + (bps1 == bps2));
        bps1.doSomething();
        
        // 5. Enum Singleton
        System.out.println("\n5. Testing Enum Singleton:");
        EnumSingleton es1 = EnumSingleton.INSTANCE;
        EnumSingleton es2 = EnumSingleton.INSTANCE;
        System.out.println("Are instances same? " + (es1 == es2));
        es1.doSomething();
        
        // Practical Example: Database Connection
        System.out.println("\nTesting Database Connection Singleton:");
        DatabaseConnection db1 = DatabaseConnection.getInstance();
        DatabaseConnection db2 = DatabaseConnection.getInstance();
        System.out.println("Are database connections same? " + (db1 == db2));
        
        // Simulate database operations
        db1.query("SELECT * FROM users");
        db2.query("INSERT INTO users (name) VALUES ('John')");
        db1.close();
    }
} 

File: javascript/designpattern/creational/SingletonPattern.js
/**
 * Singleton Design Pattern
 * 
 * Intent: Ensure a class has only one instance and provide a global point of access to it.
 * 
 * Use cases:
 * 1. Database connections
 * 2. Configuration settings
 * 3. Logger instances
 * 4. Cache managers
 * 5. State management
 * 
 * Implementation approaches:
 * 1. Object literal
 * 2. Constructor function
 * 3. ES6 Class with static instance
 * 4. Module pattern
 * 5. ES6 Module (export default)
 */

// 1. Object Literal Singleton
const ObjectLiteralSingleton = {
    instance: null,
    getInstance() {
        if (!this.instance) {
            this.instance = {
                doSomething() {
                    console.log('ObjectLiteralSingleton: Doing something...');
                }
            };
            console.log('ObjectLiteralSingleton instance created');
        }
        return this.instance;
    }
};

// 2. Constructor Function Singleton
function ConstructorSingleton() {
    // Check if instance exists
    if (ConstructorSingleton.instance) {
        return ConstructorSingleton.instance;
    }

    // Private variables and methods
    const privateVar = 'private data';

    // Public methods
    this.doSomething = function() {
        console.log('ConstructorSingleton: Doing something...');
        console.log('Accessing private variable:', privateVar);
    };

    // Store the instance
    ConstructorSingleton.instance = this;
    console.log('ConstructorSingleton instance created');
}

// 3. ES6 Class Singleton
class ClassSingleton {
    static instance = null;

    constructor() {
        if (ClassSingleton.instance) {
            return ClassSingleton.instance;
        }
        console.log('ClassSingleton instance created');
        ClassSingleton.instance = this;
    }

    doSomething() {
        console.log('ClassSingleton: Doing something...');
    }

    static getInstance() {
        if (!ClassSingleton.instance) {
            ClassSingleton.instance = new ClassSingleton();
        }
        return ClassSingleton.instance;
    }
}

// 4. Module Pattern Singleton
const ModuleSingleton = (function() {
    // Private variables and methods
    let instance = null;
    const privateVar = 'private data';

    function createInstance() {
        return {
            doSomething() {
                console.log('ModuleSingleton: Doing something...');
                console.log('Accessing private variable:', privateVar);
            }
        };
    }

    return {
        getInstance() {
            if (!instance) {
                instance = createInstance();
                console.log('ModuleSingleton instance created');
            }
            return instance;
        }
    };
})();

// 5. ES6 Module Singleton (export default)
// This would be in a separate file, but shown here for completeness
const ES6ModuleSingleton = {
    instance: null,
    getInstance() {
        if (!this.instance) {
            this.instance = {
                doSomething() {
                    console.log('ES6ModuleSingleton: Doing something...');
                }
            };
            console.log('ES6ModuleSingleton instance created');
        }
        return this.instance;
    }
};

// Example usage with a Database Connection
class DatabaseConnection {
    static instance = null;
    #connectionString;

    constructor() {
        if (DatabaseConnection.instance) {
            return DatabaseConnection.instance;
        }
        this.#connectionString = 'mysql://localhost:3306/mydb';
        console.log('Database connection established');
        DatabaseConnection.instance = this;
    }

    query(sql) {
        console.log(`Executing query: ${sql}`);
    }

    close() {
        console.log('Closing database connection');
    }

    static getInstance() {
        if (!DatabaseConnection.instance) {
            DatabaseConnection.instance = new DatabaseConnection();
        }
        return DatabaseConnection.instance;
    }
}

// Test function to demonstrate all singleton implementations
function testSingletonPattern() {
    console.log('Testing Singleton Pattern Implementations\n');

    // 1. Object Literal Singleton
    console.log('1. Testing Object Literal Singleton:');
    const obj1 = ObjectLiteralSingleton.getInstance();
    const obj2 = ObjectLiteralSingleton.getInstance();
    console.log('Are instances same?', obj1 === obj2);
    obj1.doSomething();

    // 2. Constructor Singleton
    console.log('\n2. Testing Constructor Singleton:');
    const cons1 = new ConstructorSingleton();
    const cons2 = new ConstructorSingleton();
    console.log('Are instances same?', cons1 === cons2);
    cons1.doSomething();

    // 3. Class Singleton
    console.log('\n3. Testing Class Singleton:');
    const class1 = ClassSingleton.getInstance();
    const class2 = ClassSingleton.getInstance();
    console.log('Are instances same?', class1 === class2);
    class1.doSomething();

    // 4. Module Pattern Singleton
    console.log('\n4. Testing Module Pattern Singleton:');
    const mod1 = ModuleSingleton.getInstance();
    const mod2 = ModuleSingleton.getInstance();
    console.log('Are instances same?', mod1 === mod2);
    mod1.doSomething();

    // 5. ES6 Module Singleton
    console.log('\n5. Testing ES6 Module Singleton:');
    const es61 = ES6ModuleSingleton.getInstance();
    const es62 = ES6ModuleSingleton.getInstance();
    console.log('Are instances same?', es61 === es62);
    es61.doSomething();

    // Practical Example: Database Connection
    console.log('\nTesting Database Connection Singleton:');
    const db1 = DatabaseConnection.getInstance();
    const db2 = DatabaseConnection.getInstance();
    console.log('Are database connections same?', db1 === db2);

    // Simulate database operations
    db1.query('SELECT * FROM users');
    db2.query('INSERT INTO users (name) VALUES ("John")');
    db1.close();
}

// Run the tests
testSingletonPattern(); 